Http2 to Http1.1 converter on new Http2 server API

mosesn opened the pull request:
Motivation:

http/2 and http/1.1 have similar protocols, and it's useful to be able
to implement a single server against a single interface. There's an
injection from http/1.1 messages to http/2 ones, so it makes sense to
make folks program against http/1.1 and upgrade them under the hood.

Modifications:

added a MessageToMessageCodec<Http2StreamFrame, HttpObject> which turns
every kind of Http2StreamFrame domain object into an HttpObject domain
object, and then back again on the way out.  This one is specialized for
servers, but it should be straightforward to make a symmetric one for
clients, or else extend this one.

Result:

fixes #5199, and it's now simple to make your Http2MultiplexCodec speak
Http1.1

mosesn commented:
My first nontrivial PR against netty!  Woooooooo (╯✧∇✧)╯

netkins commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,121 @@
+        }
+    }
+
+    private void encodeLastContent(LastHttpContent last, List<Object> out) {
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Make "encodeLastContent" a "static" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2325)

normanmaurer replied:
+1

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,121 @@
+public class Http2ServerDowngrader extends MessageToMessageCodec<Http2StreamFrame, HttpObject> {
+
+    @Override
+        public boolean acceptInboundMessage(Object msg) throws Exception {
```
fix formating

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Override
+        protected void decode(ChannelHandlerContext ctx, Http2StreamFrame frame, List<Object> out) throws Exception {
```
fix formating

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,121 @@
+
+    @Override
+        protected void decode(ChannelHandlerContext ctx, Http2StreamFrame frame, List<Object> out) throws Exception {
+        ReferenceCountUtil.retain(frame);
```
I think we should retain as late as possible to mimize the risk of leak data in case of exceptions.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Override
+        protected void encode(ChannelHandlerContext ctx, HttpObject obj, List<Object> out) throws Exception {
```
fix formating.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,121 @@
+
+    @Override
+        protected void encode(ChannelHandlerContext ctx, HttpObject obj, List<Object> out) throws Exception {
+        ReferenceCountUtil.retain(obj);
```
see above

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpConversionUtil.java:
```diff
@@ -251,6 +252,37 @@ public static FullHttpRequest toHttpRequest(int streamId, Http2Headers http2Head
     }
 
     /**
+     * Create a new object to contain the request data
```
Add `.` on EOL.

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2ServerDowngraderTest.java:
```diff
@@ -0,0 +1,301 @@
+        Http2ResetFrame reset = new DefaultHttp2ResetFrame(0);
+        Http2GoAwayFrame goaway = new DefaultHttp2GoAwayFrame(0);
+        ch.writeInbound(reset);
+        ch.writeInbound(goaway);
```
also assert the return value of all write calls (also in other tests).

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2ServerDowngraderTest.java:
```diff
@@ -0,0 +1,301 @@
+        assertEquals(ch.readInbound(), reset);
+        assertEquals(ch.readInbound(), goaway);
+
+        assertThat(ch.readInbound(), is(nullValue()));
```
also call channel.finish() and assert the return value

normanmaurer commented:
Just a few comments :)

@Scottmitch @nmittler @ejona86 PTAL

mosesn commented:
Thanks @normanmaurer!  All fixed up.  Looks like my editor's java formatter doesn't handle that style of annotation properly, I'll dig into it.

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,119 @@
+    @Override
+    protected void encode(ChannelHandlerContext ctx, HttpObject obj, List<Object> out) throws Exception {
+        if (obj instanceof FullHttpResponse) {
+            FullHttpResponse full = (FullHttpResponse) obj;
```
`FullHttpResponse` can also have `trailingHeaders()`

Scottmitch replied:
looks like this is covered in `encodeLastContent` ... just started looking but do you see an issue with this?

ejona86 replied:
Ah, you're right. It's fine.

ejona86 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,119 @@
+            Http2Headers headers = HttpConversionUtil.toHttp2Headers(req, true);
+            out.add(new DefaultHttp2HeadersFrame(headers, false));
+
+        } else if (obj instanceof LastHttpContent) {
```
LastHttpContent is not mutually exclusive with HttpResponse (as is seen with FullHttpResponse). Instead of if-elses, I'd suggest having separate ifs for each part of the message, and don't treat FullHttpResponse specially.

```
if (obj instanceof HttpMessage) {
  out.add(headers);
}
if (obj instanceof HttpContent) {
  if (hasContent) {
    out.add(content /* no eos */);
  }
}
if (obj instanceof LastHttpContent) {
  if (lastContent.trailingHeaders().isEmpty()) {
    out.add(emptyDataFrameWithEOS);
  } else {
    out.add(trailers);
  }
}
```

It would be possible to optimize it to avoid `emptyDataFrameWithEOS`, but I question how much that matters.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    private static void encodeLastContent(LastHttpContent last, List<Object> out) {
+        boolean needFiller = !(last instanceof FullHttpResponse) && last.trailingHeaders().isEmpty();
```
I'm not sure if we should attempt to insert "filler". Seems more expected to handle the objects as they are...

```java
LastHttpContent last = ...;
if (last.trailingHeaders().isEmpty()) {
  out.add(new dataframe(last.content(), true));
} else {
  out.add(new dataframe(last.content(), false));
  out.add(new headersframe(convert(last.trailingHeaders(), ...), true));
}
```

Scottmitch replied:
the suggestion in https://github.com/netty/netty/pull/5204#discussion_r62067945 by @ejona86 would also help mitigate this issue.

ejona86 commented:
This LGTM. I had a suggestion, but it is non-essential.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HttpConversionUtil.java:
```diff
@@ -251,6 +252,37 @@ public static FullHttpRequest toHttpRequest(int streamId, Http2Headers http2Head
+     * @return A new request object which represents headers for a chunked request
+     * @throws Http2Exception see {@link #addHttp2ToHttpHeaders(int, Http2Headers, FullHttpMessage, boolean)}
+     */
+    public static HttpRequest toHttpChunkedRequest(int streamId, Http2Headers http2Headers, boolean validateHttpHeaders)
```
consider renaming `toHttpRequest` -> `toFullHttpRequest` and renaming this `toHttpChunkedRequest` -> `toHttpRequest`. The "chunked" terminology may be confused with "chunked" encoding as defined in https://tools.ietf.org/html/rfc7230#section-4.1

Scottmitch commented:
@mosesn - Also have you signed the [ICLA](https://docs.google.com/spreadsheet/viewform?formkey=dHBjc1YzdWhsZERUQnhlSklsbG1KT1E6MQ)?

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,119 @@
+    @Override
+    protected void decode(ChannelHandlerContext ctx, Http2StreamFrame frame, List<Object> out) throws Exception {
+        if (frame instanceof Http2HeadersFrame) {
+            int id = 0; // not really the id
```
this will create problems if an exception occurs (e.g. during headers conversion). sending a stream error with stream id 0 is not allowed by the spec https://tools.ietf.org/html/rfc7540#section-6.4. We should also get the "real" stream ID to ease understanding.

@ejona86 - what is the best way to do this?

ejona86 replied:
@Scottmitch, I've got to go to a meeting, so I can't dig too deeply. But there shouldn't be much reason to need the id with the childchan API. If we really have to, we could grab it from an attribute (which currently doesn't exist) on the channel.

Scottmitch replied:
@ejona86 - is the idea that the exception should propagate through the child channel's pipeline and translated to the appropriate type of HTTP/2 frame by the http/2 codec under the hood? In the long term this seems like a good thing (assuming we can distinguish between connection/stream errors if necessary)... in the near term our Http2Exception requires the StreamId ... and if the connection stream id is used a stream error will be translated into a connection error (see [Http2Exception](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java#L124)) ... we should have some story to handle this in the near term ... wdyt?

ejona86 replied:
So I looked into this some. The value has two usages: stream exceptions and converting to non-childchan HTTP/2. Converting to non-childchan HTTP/2 doesn't seem important.

Exceptions on a stream's channel have to be handled like normal exceptions; in this case with exceptionCaught. If uncaught it just logs. Applications will commonly write exceptionCaught to close the channel, which then becomes [Http2ResetFrame(CANCEL)](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java#L413)). An application could write its own Http2ResetFrame with special error code if it wanted.

So the Http2Exceptions aren't special with h2childchan. That's true even if you're on the parent channel. And in fact it's true in the non-childchan API as well when you are later in the pipeline than the Http2ConnectionHandler; it's just that that almost never happens.

Scottmitch replied:
Yes this seems reasonable ... only concern is if a childchan uses an API which intends to throw a "Connection Error" (which would typically be converted into a GO_AWAY) ... although it seems like most of these scenarios should be on the parent channel decoding stuff before it gets to the childchan

ejona86 replied:
I would hope the Http2Exceptions would go away (or be an implementation detail) in the child chan API (for both child channel and parent channel). They stop being helpful for control flow when you have multiple handlers, since the exceptions are propagated toward the tail of the pipeline, not the head. They could still be used for notifications of failures, but it seems they couldn't trigger RST_STREAM and GOAWAY like they do today.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,119 @@
+                                                             HttpVersion.HTTP_1_1, true, true);
+                    out.add(last);
+                } else {
+                    FullHttpRequest full = HttpConversionUtil.toHttpRequest(0, headers, ctx.alloc(), true);
```
`0` -> `id` (assuming id eventually represents the "real" stream id)

Scottmitch replied:
also expose the `validateHeaders` so it can be configured by users (same as https://github.com/netty/netty/pull/5204#discussion_r62077393)

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,119 @@
+            Http2Headers headers = headersFrame.headers();
+
+            if (headersFrame.isEndStream()) {
+                if (headers.method() == null) {
```
Is the intention of this conditional to distinguish between the "headers" and the "trailers"? Perhaps a better way to do this is to check the stream state. @ejona86 - is there a way to get the stream state in the new API?

Also as a side note @nmittler I don't think we are currently enforcing this (are we?):

https://tools.ietf.org/html/rfc7540#section-8.1.2.3

> All HTTP/2 requests MUST include exactly one valid value for the
   ":method", ":scheme", and ":path" pseudo-header fields, unless it is
   a CONNECT request (Section 8.3).  An HTTP request that omits
   mandatory pseudo-header fields is malformed (Section 8.1.2.6).

Scottmitch replied:
https://tools.ietf.org/html/rfc7540#section-8.1.2.6

> Malformed requests or responses that are
   detected MUST be treated as a stream error (Section 5.4.2) of type
   PROTOCOL_ERROR.

   For malformed requests, a server MAY send an HTTP response prior to
   closing or resetting the stream.  Clients MUST NOT accept a malformed
   response.  Note that these requirements are intended to protect
   against several types of common attacks against HTTP; they are
   deliberately strict because being permissive can expose
   implementations to these vulnerabilities.

So maybe its good we don't enforce this on the server ... either it seems best to avoid using the presence/absence of these headers fields to indicate state of the stream.

ejona86 replied:
> Perhaps a better way to do this is to check the stream state. @ejona86 - is there a way to get the stream state in the new API?

No, there isn't, and currently we don't track this sort of state anywhere (even in the old API). It could be something we add, but I'd hope to not need it, especially given how hard it will be to resolve races. If the handler wasn't shareable it'd be trivial to deal with, but shareable is a laudable goal.

I don't have too much concern about relying on :method for requests and :status for responses, since they are so core to HTTP and _must_ be present in HTTP/1 during parsing. That would even behave correctly in the face of 1xx status codes where you end up having multiple sets of "headers". (For those unaware of this particular dark corner of HTTP, multiple header sets is a real thing in both [HTTP/1.1](https://tools.ietf.org/html/rfc7231#section-6.2) and [HTTP/2](https://tools.ietf.org/html/rfc7540#section-8.1). Yay!)

Another option is to add a bit to Http2HeadersFrame indicating whether it is a "header" or "trailer," but that sounds a bit icky because it has to agree with endStream.

Scottmitch replied:
> No, there isn't, and currently we don't track this sort of state anywhere (even in the old API)

In the existing API we generally pass around the stream ID or the Http2Stream object. If you have the stream ID then you can use `Http2Connection.stream(id)` to get the Http2Stream. The Http2Stream allows access to the [State](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Stream.java#L68).

> I don't have too much concern about relying on :method for requests and :status for responses, since they are so core to HTTP and must be present in HTTP/1 during parsing

Consider the following sequence of messages:

Headers (:method=get) StreamId=3 EOS=0
Data StreamId=3 EOS=0
Headers (:method=get) StreamId=3 EOS=1

This is likely a separate issue but where do we enforce that these "special" headers can only exist "exactly once" before the "data" (unless its a connect request)?

ejona86 replied:
> The Http2Stream allows access to the State.

But that state is 1) not directly usable because it has already been modified in response to the frames, 2) this logic using the state may be delayed, even on the same thread and 3) this logic using the state may be run from a different thread from where the state is maintained/changed.

For point #1, consider these new streams:

IDLE: (isn't actually tracked yet; it will [never](https://github.com/netty/netty/blob/bd6040a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L311) show up as IDLE)
OPEN: Headers (:method=get) StreamId=3 EOS=0 # These are headers
OPEN: Headers () StreamId=3 EOS=1 # These are trailers
HALF_CLOSED_REMOTE: (yes, it happens [afterward](https://github.com/netty/netty/blob/83c349f/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L324))

IDLE: (isn't actually tracked yet; it won't show up as IDLE)
HALF_CLOSED_REMOTE: Headers (:method=get) StreamId=5 EOS=1 # These are headers
(yes, this state happens [immediately](https://github.com/netty/netty/blob/bd6040a/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L315))

So the State is _never_ useful for this distinction today. That could possibly be fixed, but the child channel has its own queue. The application could simply not read from the child channel immediately and the state will be HALF_CLOSED_REMOTE in both cases. When the child channel is using a different event loop, we'd expect HALF_CLOSED_REMOTE in both cases.

> This is likely a separate issue but where do we enforce that these "special" headers can only exist "exactly once" before the "data" (unless its a connect request)?

It looks like we should be prohibiting that (still taking into account 1xx):

> Pseudo-header fields are only valid in the context in which they are defined. Pseudo-header fields defined for requests MUST NOT appear in responses; pseudo-header fields defined for responses MUST NOT appear in requests. Pseudo-header fields MUST NOT appear in trailers. Endpoints MUST treat a request or response that contains undefined or invalid pseudo-header fields as malformed (Section 8.1.2.6).

Scottmitch replied:
Agreed the state as it exists today won't be very useful for this purpose

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,119 @@
+
+            if (headersFrame.isEndStream()) {
+                if (headers.method() == null) {
+                    LastHttpContent last = new DefaultLastHttpContent(Unpooled.EMPTY_BUFFER, true);
```
typically the `validateHeaders` is exposed and configurable by the user. validating headers can add non-trivial overhead that some users decide to forgo.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,119 @@
+                    out.add(full);
+                }
+            } else {
+                out.add(HttpConversionUtil.toHttpChunkedRequest(id, headersFrame.headers(), true));
```
expose the `validateHeaders` so it can be configured by users (same as https://github.com/netty/netty/pull/5204#discussion_r62077393)

mosesn commented:
I've simplified the request path, renamed toHttpRequest/toFullHttpRequest, and added validateHeaders.

Some places I need more feedback:
A.  What is the right default for validateHeaders?  on or off?
B.  Is there a better way to distinguish headers and trailers?  I would rather not have to maintain stream state, since right now it's stateless, and can be sharable.

I've added a new commit so that it's easier to see the diff, but I can squash it once we're happy with the state it's in.

Please take a look!

mosesn commented:
OK, just wanted to follow up.  It doesn't sound like any new feedback is coming in, so is this ready to merge?  If it is, then I'll squash and push again.

Scottmitch commented:
@mosesn - sorry for the delay ... been busy this week ... will look tomorrow.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    public Http2ServerDowngrader() {
+        this.validateHeaders = true;
```
nit: `this(true)`

mosesn replied:
done, thanks!

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,126 @@
+            boolean noMoreFrames = false;
+            if (obj instanceof FullHttpResponse) {
+                FullHttpResponse full = (FullHttpResponse) obj;
+                noMoreFrames = full.content().readableBytes() == 0 && full.trailingHeaders().isEmpty();
```
`full.content().readableBytes() == 0` -> `!full.content().isReadable()`

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,126 @@
+        ReferenceCountUtil.retain(frame);
+    }
+
+    private static void encodeLastContent(LastHttpContent last, List<Object> out, boolean validateHeaders) {
```
why not make non-static and just use `validateHeaders` as a member variable instead of as a parameter?

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    private static void encodeLastContent(LastHttpContent last, List<Object> out, boolean validateHeaders) {
+        boolean needFiller = !(last instanceof FullHttpResponse) && last.trailingHeaders().isEmpty();
```
its not clear why we should try to simulate the user sending a data frame to make the encoding work. Seems more intuitive to just encode what the user provides and if their logic is flawed they should fix it. Can we just mimic the logic in [HttpToHttp2ConnectionHandler.write](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java) or improve the logic their if it is not good enough?  It would be nice to share code or at least use consistent logic if possible.

mosesn replied:
We need to handle the case where a user sends a `LastHttpContent.EMPTY_LAST_CONTENT` in order to close their streamed message, but there's no need to send a message otherwise.  I don't think this is incorrect behavior.

It looks like in HttpToHttp2ConnectionHandler, it looks like it sends a data frame regardless of whether there's content or not, which seems unnecessary.  I can clean it up, do you mind if we push to a subsequent review though?

Scottmitch replied:
>  I don't think this is incorrect behavior.

Agreed. Thanks for clarifying.

>  I can clean it up, do you mind if we push to a subsequent review though?

Followup PR sgtm thanks!

mosesn commented:
@Scottmitch Thanks for the advice, I've updated the branch.  PTAL!

Scottmitch commented:
@mosesn - LGTM. Please squash and I will cherry-pick.

normanmaurer commented:
@Scottmitch @mosesn wooohooo :)

mosesn commented:
@Scottmitch thanks for the thorough review, I've squashed my commit ✨

Scottmitch commented:
Cherry-picked 4.1 (9ce84dc)

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2ServerDowngraderTest.java:
```diff
@@ -0,0 +1,317 @@
+        assertThat(request.protocolVersion(), is(HttpVersion.HTTP_1_1));
+        assertThat(request.content().readableBytes(), is(0));
+        assertTrue(request.trailingHeaders().isEmpty());
+
```
@mosesn - I just merged this an noticed a leak in the build log. can you add another commit (or pr) where these `request` is released in a finally block for all these tests?

mosesn replied:
yep, how can I check if I'm leak-free?

Scottmitch replied:
The best thing we have ATM is [-Dio.netty.leakDetectionLevel=paranoid](http://netty.io/wiki/reference-counted-objects.html#leak-detection-levels).

Scottmitch replied:
thanks :)

normanmaurer replied:
@mosesn run a few times with "-Pleak" and see if you see any `LEAK:` logs in the console

Scottmitch replied:
note that you can also use [releaseLater](http://netty.io/wiki/reference-counted-objects.html#fixing-leaks-in-unit-tests)

mosesn replied:
I've run it a few times, but can't repro locally.  Here's what I'm doing:

```bash
mvn test -pl codec-http2 -Dtest=Http2ServerDowngraderTest -Pleak
```

Is there anything I'm missing?  Is there a way I can see the failed build, so I can fix based on that?

Scottmitch replied:
@mosesn - the build won't fail locally, but you should see log statements that look like this https://garage.netty.io/teamcity/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=8253#_focus=18593

mosesn replied:
Weird, I don't see that failure locally.  Do you see it consistently?  Do I need to change a log level somewhere?

mosesn replied:
Not sure why, but it just started working locally.  Working on a fix now–looks like it's just for one test.

mosesn replied:
I made a PR, but I'm not sure if this is the right fix: #5230.
